### PR TITLE
FFT: disable dimension wrapping for scalar tensors

### LINKED ISSUE #88985: `torch.fft.hfft` Trigger RuntimeError under UndefinedBehaviorSanitizer
### 🐛 Describe the bug

The implementation of `torch.fft.rfft` contains undefined behavior caught by UB sanitizer.
Test:
```python
import torch

def test():
    arg_1 = torch.rand([], dtype=torch.float64).clone()
    res = torch.fft.hfft(arg_1,1)

test()
```
Error log:
```
pytorch/c10/util/SmallVector.h:340:12: runtime error: addition of unsigned offset to 0x7f28c4183090 overflowed to 0x7f28c4183088
```

### Versions

```
PyTorch version: 1.14.0a0+git6e5f736
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 11.1.0-6
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.9.12 (main, Apr  5 2022, 06:56:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-52-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090

Nvidia driver version: 515.65.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.5
[pip3] numpydoc==1.2
[pip3] torch==1.14.0a0+git6e5f736
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.21.5           py39he7a7128_1  
[conda] numpy-base                1.21.5           py39hf524024_1  
[conda] numpydoc                  1.2                pyhd3eb1b0_0  
[conda] torch                     1.14.0a0+git6e5f736          pypi_0    pypi
```

cc @mruberry @peterbell10

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89234

Fixes #88985

By default, `maybe_wrap_dim` allows through `dim=0` or `dim=-1`
for scalar tensors which leads to an invalid dimension being used to
index into `tensor.sizes()` as in the code sample from the issue.

cc @mruberry

### CODE REVIEW

**Comment by mruberry** [comment]
@peterbell10 could you elaborate on this change with a before and after examples? Are there any backwards compatibility concerns here?

**Comment by peterbell10** [comment]
Take `fft` as an example. Before this PR I get this
```
In [1]: import torch
   ...: a = torch.rand(())
   ...: torch.fft.fft(a)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [1], in <cell line: 3>()
      1 import torch
      2 a = torch.rand(())
----> 3 torch.fft.fft(a)

RuntimeError: Invalid number of data points (0) specified
```

`fft` must transform exactly one dimension, but here it's allowed `dim=-1` (the default for `fft`) on a zero-dimensional tensor. The error in this case makes no sense  because it's reading from uninitialized memory that happens to be a zero but in general might be garbage.

After this change, the error message becomes 
```
IndexError: Dimension specified as -1 but tensor has no dimensions
```

There is no backward compatibility implication because the previous zero-dim behavior depended on undefined behavior and was purely a bug.

File: aten/src/ATen/WrapDimUtils.h [hunk 1]
```diff
@@ -43,9 +43,19 @@ inline int64_t maybe_wrap_dim(
 inline void maybe_wrap_dims_n(
     int64_t* dims,
     int64_t ndims,
-    int64_t dim_post_expr) {
+    int64_t dim_post_expr,
+    bool wrap_scalars = true) {
```
Comment (mruberry) [inline aten/src/ATen/WrapDimUtils.h:47]: Let's extend the comment describing this function to include a description of each parameter and what they do

File: aten/src/ATen/WrapDimUtils.h [hunk 2]
```diff
@@ -70,8 +80,12 @@ inline void maybe_wrap_dims_n(
 // Wrap each dim in a contiguous container, taking dim_post_expr as the true
 // number of dimensions E.g. could also be std::array or c10::SmallVector
 template <typename Container>
-inline void maybe_wrap_dims(Container& dims, int64_t dim_post_expr) {
-  return maybe_wrap_dims_n(dims.data(), dims.size(), dim_post_expr);
+inline void maybe_wrap_dims(
```
Comment (mruberry) [inline aten/src/ATen/WrapDimUtils.h:83]: Improved comment needed here, too

File: torch/testing/_internal/opinfo/definitions/fft.py [hunk 3]
```diff
@@ -65,6 +67,24 @@ def __init__(
         super().__init__(**ukwargs)
 
 
+def error_inputs_fft(op_info, device, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=torch.float32)
```
Comment (mruberry) [inline torch/testing/_internal/opinfo/definitions/fft.py:71]: Add a comment describing the error tested for both these `ErrorInput`s

**Review by mruberry — APPROVED** [review]
Cool! But see requests for some additional comments

**Pull request closed without merging**